pkg/migrate/attributemigration.go

```diff
@@ -0,0 +1,301 @@
+	resources, err := m.client.GetAPIV1ResourceInstancesWithPageSize(query, resourceURL, 100)
+	if err != nil {
+		return err
+	}
+
+	items := make([]item, 0)
+
+	for _, ri := range resources {
+		item := updateAttrs(ri)
+		items = append(items, item)
+	}
+
+	wg := &sync.WaitGroup{}
+	errCh := make(chan error, len(items))
+
+	for _, item := range items {
+		if !item.update {
+			continue
+		}
+
+		wg.Add(1)
+
+		go func(ri *v1.ResourceInstance) {
+			defer wg.Done()
+
+			url := fmt.Sprintf("%s/%s", resourceURL, ri.Name)
+			err := m.updateRI(url, ri)
+			errCh <- err
+		}(item.ri)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
+}
+
+// updateRI updates the resource, and the sub resource
+func (m *AttributeMigration) updateRI(url string, ri *v1.ResourceInstance) error {
+	_, err := m.client.UpdateAPIV1ResourceInstance(url, ri)
+	if err != nil {
+		return err
+	}
+
+	return m.createSubResource(ri)
+}
+
+func (m *AttributeMigration) getRI(url string) (*v1.ResourceInstance, error) {
+	response, err := m.client.ExecuteAPI(api.GET, url, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error while retrieving ResourceInstance: %s", err)
+	}
+
+	resourceInstance := &v1.ResourceInstance{}
+	err = json.Unmarshal(response, &resourceInstance)
+	return resourceInstance, err
+}
+
+func (m *AttributeMigration) createSubResource(ri *v1.ResourceInstance) error {
+	plural, err := getPlural(ri.Kind)
+	if err != nil {
+		return err
+	}
+
+	err = m.client.CreateSubResourceScoped(
+		mv1a.EnvironmentResourceName,
+		m.cfg.GetEnvironmentName(),
+		plural,
+		ri.Name,
+		ri.Group,
+		ri.APIVersion,
+		ri.SubResources,
+	)
+
+	return err
+}
+
+func getPlural(kind string) (string, error) {
+	switch kind {
+	case mv1a.APIServiceGVK().Kind:
+		return mv1a.APIServiceResourceName, nil
+	case mv1a.APIServiceRevisionGVK().Kind:
+		return mv1a.APIServiceRevisionResourceName, nil
+	case mv1a.APIServiceInstanceGVK().Kind:
+		return mv1a.APIServiceInstanceResourceName, nil
+	case mv1a.ConsumerInstanceGVK().Kind:
+		return mv1a.ConsumerInstanceResourceName, nil
+	default:
+		return "", fmt.Errorf("cannot get plural for %s", kind)
+	}
+}
+
+func updateAttrs(ri *v1.ResourceInstance) item {
+	details := util.GetAgentDetails(ri)
+	if details == nil {
+		details = make(map[string]interface{})
+	}
+
+	item := item{
+		ri:     ri,
+		update: false,
+	}
+
+	for _, attr := range oldAttrs {
+		if _, ok := ri.Attributes[attr]; ok {
+			details[attr] = ri.Attributes[attr]
+			delete(ri.Attributes, attr)
+			item.update = true
+		}
+	}
+
+	for _, reg := range regexes {
+		for attr := range ri.Attributes {
+			if ok, _ := regexp.MatchString(reg, attr); ok {
+				details[attr] = ri.Attributes[attr]
+				delete(ri.Attributes, attr)
+				item.update = true
+			}
+		}
+	}
+
+	util.SetAgentDetails(ri, details)
+
+	return item
+}
+
+func queryFunc(name string) string {
+	return fmt.Sprintf("metadata.references.name==%s", name)
```

Comment:
@vivekschauhan @jcollins-axway I found a bug. This query does not work for api service instances. It is the only resource that does not have a reference to the API Service. It only has a reference to the revision it was made for. The revision and consumer instances both have a reference to the API Service. Can we update the discovery process to include a ref to the API Service when creating the APIServiceInstance?